File: frontend/src/types/assets.ts
export interface Tag {
  id: string;
  label: string;
}

export interface ImageAsset {
  id: string;
  label: string;
  artist: string | null;
  src: string;
  thumb_src: string | null;
  url: string;
  thumb_url: string | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
}

export interface AudioAsset {
  id: string;
  label: string;
  artist: string | null;
  src: string;
  url: string;
  duration: number | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
}

export interface VideoAsset {
  id: string;
  label: string;
  artist: string | null;
  src: string;
  thumb_src: string | null;
  url: string;
  thumb_url: string | null;
  duration: number | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
}

export type AnyAsset = ImageAsset | AudioAsset | VideoAsset;
export type AssetType = "image" | "audio" | "video";

export interface OrphanedFile {
  key: string;
}

export interface BrokenAsset {
  id: string;
  label: string;
  src: string;
  type: AssetType;
}

export interface ReconcileResult {
  orphaned_files: OrphanedFile[];
  broken_assets: BrokenAsset[];
}

File: backend/app/models/assets.py
"""Asset table models for images, audio, and video files stored on the CDN."""

from datetime import datetime
from typing import Optional
from uuid import UUID, uuid4

from sqlalchemy import func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ImageAsset(Base):
    __tablename__ = "image_asset"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    label: Mapped[str]
    artist: Mapped[Optional[str]] = mapped_column(default=None)
    src: Mapped[str]
    thumb_src: Mapped[Optional[str]] = mapped_column(default=None)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())

    tags: Mapped[list["Tag"]] = relationship(secondary="image_asset_tag")


class AudioAsset(Base):
    """Used for both ambience sounds and music tracks."""

    __tablename__ = "audio_asset"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    label: Mapped[str]
    artist: Mapped[Optional[str]] = mapped_column(default=None)
    src: Mapped[str]
    duration: Mapped[Optional[float]] = mapped_column(default=None)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())

    tags: Mapped[list["Tag"]] = relationship(secondary="audio_asset_tag")


class VideoAsset(Base):
    __tablename__ = "video_asset"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    label: Mapped[str]
    artist: Mapped[Optional[str]] = mapped_column(default=None)
    src: Mapped[str]
    duration: Mapped[Optional[float]] = mapped_column(default=None)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())

    tags: Mapped[list["Tag"]] = relationship(secondary="video_asset_tag")


from app.models.tags import Tag  # noqa: E402
